osquery/tables/system/darwin/nvram.cpp

```diff
@@ -19,47 +19,93 @@
-    return;
-  }
-
-  // Variable name is the dictionary key.
-  Row nvram_row;
-  nvram_row["name"] = stringFromCFString((CFStringRef)key);
+const std::string kIODTOptionsPath = "IODeviceTree:/options";
 
+Status stringFromNVRAM(const void* value,
+                       std::string& type_name,
+                       std::string& output) {
   // Variable type will be defined by the CF type.
   CFTypeID type_id = CFGetTypeID(value);
   CFStringRef type_description = CFCopyTypeIDDescription(type_id);
-  nvram_row["type"] = stringFromCFString(type_description);
+  type_name = stringFromCFString(type_description);
+
   CFRelease(type_description);
 
   // Based on the type, get a texual representation of the variable.
-  std::string value_string;
   if (type_id == CFBooleanGetTypeID()) {
-    value_string = (CFBooleanGetValue((CFBooleanRef)value)) ? "true" : "false";
+    output = (CFBooleanGetValue((CFBooleanRef)value)) ? "true" : "false";
   } else if (type_id == CFNumberGetTypeID()) {
-    value_string = stringFromCFNumber((CFDataRef)value);
+    output = stringFromCFNumber((CFDataRef)value);
   } else if (type_id == CFStringGetTypeID()) {
-    value_string = stringFromCFString((CFStringRef)value);
+    output = stringFromCFString((CFStringRef)value);
   } else if (type_id == CFDataGetTypeID()) {
-    value_string = stringFromCFData((CFDataRef)value);
+    output = stringFromCFData((CFDataRef)value);
   } else {
     // Unknown result type, do not attempt to decode/format.
-    value_string = "<INVALID>";
+    return Status::failure("Unknown variable type");
   }
+  return Status::success();
+}
 
-  // Finally, add the variable value to the row.
-  nvram_row["value"] = value_string;
-  ((QueryData *)results)->push_back(nvram_row);
+void genVariable(const void* key, const void* value, void* results) {
+  if (key == nullptr || value == nullptr || results == nullptr) {
+    // Paranoia: don't expect the callback application to yield nullptrs.
+    return;
+  }
+
+  // Variable name is the dictionary key.
+  Row r;
+  auto name = stringFromCFString((CFStringRef)key);
+
+  std::string type_name;
+  std::string value_string;
+  auto status = stringFromNVRAM(value, type_name, value_string);
+  if (!status.ok()) {
+    VLOG(1) << "Failed to convert NVRAM variable: " << name;
+  } else {
+    r["value"] = std::move(value_string);
+  }
+
+  r["name"] = std::move(name);
+  r["type"] = std::move(type_name);
+  ((QueryData*)results)->push_back(r);
 }
 
-QueryData genNVRAM(QueryContext &context) {
-  QueryData results;
+void genSingleVariable(const io_registry_entry_t& options,
+                       const std::string& key,
+                       QueryData& results) {
+  auto name = CFStringCreateWithCString(
+      kCFAllocatorDefault, key.c_str(), kCFStringEncodingUTF8);
+  if (name == nullptr) {
+    VLOG(1) << "Cannot create CFString for NVRAM name";
+    return;
+  }
+
+  auto value = IORegistryEntryCreateCFProperty(options, name, 0, 0);
```

Comment:
Should we not prefer `kCFAllocatorDefault` as the third parameter to `IORegistryEntryCreateCFProperty`? While `0` is equivalent, having `kCFAllocatorDefault` is more readable and in line with the rest of IOKit code.